Fit benchmarking output results to HTML

simonfernandes opened the pull request:
Allows group results to additionally be outputted to an HTML file.

When the flag `save_to_file` is set to true, the reST data is parsed using docutils and placed in an HTML file, along with the reST data .txt files. These two files per group are placed in the same location (the script directory).

Summary tables are not included. The tables appear to be malformed (table data not fully enclosed by reST grid table ASCII) and they appear to use a role (`:ref:`) only supported by Sphinx.

**To test:**
Ensure that individual HTML files are produced for all groups specified, in the correct directory.

Fixes #18422 

<!-- RELEASE NOTES
Either edit the file in docs/source/release/... and it will be in your pull request or state
*Does not need to be in the release notes.*
-->

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

Anders-Markvardsen commented:
When I run it get the error message: 

IOError: [Errno 2] No such file or directory: 'color_definitions.txt'
  at line 55 in 'C:/Backup/Backup_folder1/work/code/Mantid/fitting/FitMinimizerTesting/runScripts.py'
  caused by line 87 in 'C:/Backup/Backup_folder1/work/code/Mantid/git/mantid/scripts/CompareFitMinimizers\results_output.py'

simonfernandes commented:
@Anders-Markvardsen Sorry, my mistake, I forgot that I had moved the file into the script directory for testing. I have fixed the location to it now. There is an additional problem in getting the colours to show now:

The `color_defintions.txt` file is located in `docs/source/concepts/minimizers_comparison`, but looks for `color_definitions.js` in the directory below it. There is also a slight difference in the name of the file it looks for.

I will edit the above code snippet change to link to the correct file and directory.

Anders-Markvardsen commented:
Can you have the script write out where it saves the file

Anders-Markvardsen commented on scripts/CompareFitMinimizers/results_output.py:
```diff
@@ -74,14 +80,23 @@ def print_group_results_tables(minimizers, results_per_test, problems_obj, group
         header += "\n\n"
         print(header)
         print (tbl_acc_indiv)
+        rel_col_defs = '../../docs/source/concepts/minimizers_comparison/color_definitions.txt'
```
I would be better if a modifiable copy of this file was in scripts\CompareFitMinimizers or subfolder of this directory. This is to decouple the concern of generating the concept pages from that of generating the fit benchmarking tables. Also together with the .txt file need a copy of the accompanying docs\source\_static\color_ranking_definitions.js + some container rst to wrap this color scheme with the rst tables for html generation

Anders-Markvardsen commented on scripts/CompareFitMinimizers/results_output.py:
```diff
@@ -74,14 +80,23 @@ def print_group_results_tables(minimizers, results_per_test, problems_obj, group
 
-        # optionally save the above table to file
+        # optionally save the above table to a .txt file and a .html file
         if save_to_file:
```
The number of lines within the save_to_file I would say now justify a new method called something like save_table_to_file

Anders-Markvardsen commented on scripts/CompareFitMinimizers/results_output.py:
```diff
@@ -120,6 +126,22 @@ def print_group_results_tables(minimizers, results_per_test, problems_obj, group
         print(tbl_runtime_summary)
 
 
+def save_table_to_file(table_data, errors, group_name, metric_type, file_extension):
```
Looks good with refactoring this code into a new method. Could you add documentation please to this method please including for its arguments

Anders-Markvardsen commented on scripts/CompareFitMinimizers/Definitions/color_definitions.txt:
```diff
@@ -0,0 +1,22 @@
+      of the parents of the text/numbers that go in some table cells (as span elements), so that the cells themselves
+      (parents) are assigned a certain style class (background color)
+
+.. raw:: html
```
Could you investigate if the content of *.js file could be included in the ..raw:: html? If this can be done it would mean the html files can be generated and located anywhere

Anders-Markvardsen commented:
@simonfernandes fixed conflict, which ross doing some maintenance on the scripts to try numpy before scipy.stats

Anders-Markvardsen commented:
Tested with single and multiple minimizers and works for individual tables. The unrelated issue for summary tables will be handled in separate issue.

:shipit:
